mozilla_django_oidc/utils.py

```diff
@@ -54,16 +55,75 @@ def is_authenticated(user):
     return user.is_authenticated
 
 
-def add_state_and_nonce_to_session(request, state, params):
+def base64_url_encode(bytes_like_obj):
+    """Return a URL-Safe, base64 encoded version of bytes_like_obj
+
+    Implements base64urlencode as described in
+    https://datatracker.ietf.org/doc/html/rfc7636#appendix-A
+    """
+
+    s = josepy.b64.b64encode(bytes_like_obj).decode("ascii")  # base64 encode
+    # the josepy base64 encoder (strips '='s padding) automatically
+    s = s.replace("+", "-")  # 62nd char of encoding
+    s = s.replace("/", "_")  # 63rd char of encoding
+
+    return s
+
+
+def base64_url_decode(string_like_obj):
+    """Return the bytes encoded in a URL-Safe, base64 encoded string
+    Implements inverse of base64urlencode as described in
+    https://datatracker.ietf.org/doc/html/rfc7636#appendix-A
+    This function is not used by the OpenID client; it's just for testing PKCE related functions.
+    """
+    s = string_like_obj
+
+    s = s.replace("_", "/")  # 63rd char of encoding
+    s = s.replace("-", "+")  # 62nd char of encoding
+    b = josepy.b64.b64decode(s)  # josepy base64 encoder (decodes without '='s padding)
+
+    return b
+
+
+def generate_code_challenge(code_verifier, method):
+    """Return a code_challege, which proves knowledge of the code_verifier.
+    The code challenge is generated according to method which must be one
+    of the methods defined in https://datatracker.ietf.org/doc/html/rfc7636#section-4.2:
+    - plain:
+      code_challenge = code_verifier
+    - S256:
+      code_challenge = BASE64URL-ENCODE(SHA256(ASCII(code_verifier)))
     """
-    Stores the `state` and `nonce` parameters in a session dictionary including the time when it
-    was added. The dictionary can contain multiple state/nonce combinations to allow parallel
-    logins with multiple browser sessions.
-    To keep the session space to a reasonable size, the dictionary is kept at 50 state/nonce
-    combinations maximum.
+
+    if method == "plain":
+        return code_verifier
+
+    elif method == "S256":
+        return base64_url_encode(sha256(code_verifier.encode("ascii")).digest())
+
+    else:
+        raise ValueError("code challenge method must be 'plain' or 'S256'.")
+
+
+def add_state_and_verifier_and_nonce_to_session(
+    request, state, params, code_verifier=None
```

Comment:
Nit. Why not look for `code_verifier` in the `params`, and use `code_verifier = params.get("code_verifier")` to get its value?

Reply:
Valid point but let's optimize one this is out. It's already waiting for a while now and I would prefer to get this out sooner than later